src/Microsoft.ApplicationInsights.AspNetCore/DiagnosticListeners/Implementation/CorrelationIdLookupHelper.cs

```diff
@@ -0,0 +1,199 @@
+        /// Initializes a new instance of the <see cref="CorrelationIdLookupHelper" /> class mostly to be used by the test classes to provide an override for fetching appId logic.
+        /// </summary>
+        /// <param name="appIdProviderMethod">The delegate to be called to fetch the appId.</param>
+        public CorrelationIdLookupHelper(Func<string, Task<string>> appIdProviderMethod)
+        {
+            if (appIdProviderMethod == null)
+            {
+                throw new ArgumentNullException(nameof(appIdProviderMethod));
+            }
+
+            this.provideAppId = appIdProviderMethod;
+        }
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CorrelationIdLookupHelper" /> class.
+        /// </summary>
+        /// <param name="configurationProvider">A delegate that provides telemetry configuration, which is to be used to fetch endpoint address.</param>
+        public CorrelationIdLookupHelper(Func<TelemetryConfiguration> configurationProvider)
+        {
+            this.configurationProvider = configurationProvider;
+            this.provideAppId = this.FetchAppIdFromService;
+        }
+
+        /// <summary>
+        /// Retrieves the correlation id corresponding to a given instrumentation key.
+        /// </summary>
+        /// <param name="instrumentationKey">Instrumentation key string.</param>
+        /// <param name="correlationId">AppId corresponding to the provided instrumentation key.</param>
+        /// <returns>true if correlationId was successfully retrieved; false otherwise.</returns>
+        public bool TryGetXComponentCorrelationId(string instrumentationKey, out string correlationId)
+        {
+            if (string.IsNullOrEmpty(instrumentationKey))
+            {
+                // This method cannot throw - it's a Try... method. We cannot proceed any further.
+                correlationId = string.Empty;
+                return false;
+            }
+
+            var found = this.knownCorrelationIds.TryGetValue(instrumentationKey, out correlationId);
+
+            if (found)
+            {
+                return true;
+            }
+            else
+            {
+                // Simplistic cleanup to guard against this becoming a memory hog.
+                if (this.knownCorrelationIds.Keys.Count >= MAXSIZE)
+                {
+                    this.knownCorrelationIds.Clear();
+                }
+
+                try
+                {
+
+                    // We wait for <getAppIdTimeout> seconds (which is 0 at this point) to retrieve the appId. If retrieved during that time, we return success setting the correlation id.
+                    // If we are still waiting on the result beyond the timeout - for this particular call we return the failure but queue a task continuation for it to be cached for next time.
+                    Task<string> getAppIdTask = this.provideAppId(instrumentationKey.ToLowerInvariant());
```

Comment:
I see the value here to prevent task querying the same app id being scheduled. Code updated in the next iteration.
Thanks!